django/db/migrations/state.py

```diff
@@ -205,8 +234,29 @@ def alter_field(self, app_label, model_name, name, field, preserve_default):
             field.default = NOT_PROVIDED
         else:
             field = field
-        model_state = self.models[app_label, model_name]
-        model_state.fields[name] = field
+        model_key = (app_label, model_name)
+        model_state = self.models[model_key]
+        fields = model_state.fields
+        if self._relations is not None:
+            old_remote_field = fields[name].remote_field
+            new_remote_field = field.remote_field
+            if old_remote_field and new_remote_field:
+                if old_remote_field.model != new_remote_field.model:
+                    old_field = fields.pop(name)
+                    self.resolve_model_field_relations(model_key, name, old_field)
+                    fields[name] = field
+                    self.resolve_model_field_relations(model_key, name, field)
+            else:
+                if old_remote_field:
+                    old_field = fields.pop(name)
+                    self.resolve_model_field_relations(model_key, name, old_field)
+                    fields[name] = field
+                elif new_remote_field:
+                    fields[name] = field
+                    self.resolve_model_field_relations(model_key, name, field)
+        else:
```

Comment:
This would be cleaner if you did the following before handling the `self._relations is not None` case:

```python
if self._relations is None:
    fields[name] = field
    return
```

Then you don't need to indent after.

Reply:
That won't be feasible as in other cases, we have to pop the old field before setting the new field.

Reply:
Oh, when I was reading before, I thought the function ended there. It could still help readability to do the `if self._relations is None:` case first, since there are less "nots" to look at.

Reply:
Simplified.